### How to increase sales in a certain area(especially outside Java)?
This question has been asked every month by managers and c levels. Retail is a complex business, and some areas have more product types compared to others. For example:
- Area A, sells 10 different types of shampoo and 2 different types of chips
- Area B, sells 3 different types of shampoo and 5 different types of chips

Through experimentation by adding a certain product type to an area, sales performance can be increased.
Determing which product to be placed in a certain area(especially outside Java) is crucial because of difference pricing, difference promotions, etc.

However, the purpose of this notebook is to learn from the transaction behavior of an area with top sales of a product that we want to expand to other areas by using Apriori:
- When a user bought product X, what product that is likely to be bought together?

Here's the approach:
1. Get apriori of an area or a few with top sales(for example fruit, this is top sales in branch Jakarta and "what if we try to sell it in other area?")
2. Learnt its pair product
3. then search for area that we want to improve the sales + not selling the fruit
4. check their top sales if match the second point(pair product) then we can consider this area

In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [2]:
df = pd.read_excel('/content/dummy_data.xlsx')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6259 entries, 0 to 6258
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   order_id          6259 non-null   object
 1   product_category  6259 non-null   object
dtypes: object(2)
memory usage: 97.9+ KB


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [6]:
df_encoded = pd.get_dummies(df['product_category']).groupby(df['order_id']).apply(max)

min_support = 0.01
frequent_itemsets = apriori(df_encoded, min_support=min_support, use_colnames=True)

min_confidence = 0.01
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)

rules["lift"] = rules["lift"].round(2)


rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents                           consequents   support  \
0             (ULTRA MILK)                         (AIR FRESHER)  0.011100   
1            (AIR FRESHER)                          (ULTRA MILK)  0.011100   
2                   (AQUA)                          (GREENFIELD)  0.010595   
3             (GREENFIELD)                                (AQUA)  0.010595   
4             (ULTRA MILK)                                (AQUA)  0.032291   
..                     ...                                   ...       ...   
849  (ULTRA MILK, INDOMIE)                (SUGAR, MINYAK GORENG)  0.011100   
850                (SUGAR)  (ULTRA MILK, MINYAK GORENG, INDOMIE)  0.011100   
851        (MINYAK GORENG)          (SUGAR, INDOMIE, ULTRA MILK)  0.011100   
852              (INDOMIE)    (SUGAR, MINYAK GORENG, ULTRA MILK)  0.011100   
853           (ULTRA MILK)       (SUGAR, MINYAK GORENG, INDOMIE)  0.011100   

     confidence  lift  
0      0.011117  1.00  
1      1.000000  1.00  
2      0.328125  1.97  
3      0.063636  1.97  
4      0.032340  1.00  
..          ...   ...  
849    0.068111  4.22  
850    0.207547  5.96  
851    0.134146  5.02  
852    0.067485  4.18  
853    0.011117  1.00  

[854 rows x 5 columns]

In [8]:
filtered_rules = rules[rules['consequents'].apply(lambda x: 'CHIP CHIP' in str(x))]
filtered_rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']].sort_values(by='lift', ascending=False).head(20)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents              consequents   support  confidence  \
22                  (BISCUIT)              (CHIP CHIP)  0.010595    0.221053   
276     (ULTRA MILK, BISCUIT)              (CHIP CHIP)  0.010595    0.221053   
280                 (BISCUIT)  (ULTRA MILK, CHIP CHIP)  0.010595    0.221053   
400                 (COOKIES)  (ULTRA MILK, CHIP CHIP)  0.012109    0.196721   
396     (ULTRA MILK, COOKIES)              (CHIP CHIP)  0.012109    0.196721   
64                  (COOKIES)              (CHIP CHIP)  0.012109    0.196721   
423                  (POTATO)  (ULTRA MILK, CHIP CHIP)  0.013623    0.188811   
420      (POTATO, ULTRA MILK)              (CHIP CHIP)  0.013623    0.188811   
72                   (POTATO)              (CHIP CHIP)  0.013623    0.188811   
312       (BREAD, ULTRA MILK)              (CHIP CHIP)  0.015641    0.128631   
315                   (BREAD)  (ULTRA MILK, CHIP CHIP)  0.015641    0.128631   
36                    (BREAD)              (CHIP CHIP)  0.015641    0.128631   
408     (ULTRA MILK, INDOMIE)              (CHIP CHIP)  0.016650    0.102167   
68                  (INDOMIE)              (CHIP CHIP)  0.016650    0.101227   
412                 (INDOMIE)  (ULTRA MILK, CHIP CHIP)  0.016650    0.101227   
415        (ULTRA MILK, LOGI)              (CHIP CHIP)  0.011604    0.085821   
71                     (LOGI)              (CHIP CHIP)  0.011604    0.085502   
419                    (LOGI)  (ULTRA MILK, CHIP CHIP)  0.011604    0.085502   
406              (GREENFIELD)  (ULTRA MILK, CHIP CHIP)  0.014127    0.084848   
402  (ULTRA MILK, GREENFIELD)              (CHIP CHIP)  0.014127    0.084848   

     lift  
22   4.47  
276  4.47  
280  4.47  
400  3.98  
396  3.98  
64   3.98  
423  3.82  
420  3.82  
72   3.82  
312  2.60  
315  2.60  
36   2.60  
408  2.07  
68   2.05  
412  2.05  
415  1.74  
71   1.73  
419  1.73  
406  1.72  
402  1.72

By looking at the output, on the first row we can assume that:
- Antecedent: If a customer buys "BISCUIT",
- Consequent: Then there's a 22.1% chance they will also buy "(CHIP CHIP)".
- Support: Only about 1.06% of all transactions contain both "BISCUIT" and "(CHIP CHIP)".
- Confidence: Out of those who bought "BISCUIT," 22.1% also bought "(CHIP CHIP)".
- Lift: The likelihood of buying "(CHIP CHIP)" is 4.47 times higher when "BISCUIT" is in the basket compared to random chance.

Reference:
- https://towardsdatascience.com/data-mining-market-basket-analysis-with-apriori-algorithm-970ff256a92c
- https://www.datacamp.com/blog/5-ways-use-data-science-marketing